### PR TITLE
Make `sleep` work with isolation enabled

### PR DESCRIPTION
**Pull request by pvdrz**
Implement a virtual monotone clock that can be used to track time while isolation is enabled. This virtual clock keeps an internal nanoseconds counter that will be increased by a fixed amount at the end of every basic block.

When a process sleeps, this clock will return immediately and increase the counter by the interval the process was supposed to sleep. Making miri execution faster than native code :trollface:.

cc @RalfJung @saethlin @JakobDegen

### CODE REVIEW

File: tests/pass/shims/time-with-isolation.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+//@compile-flags: -Zmiri-disable-isolation
+
+use std::time::{Duration, Instant};
+
+fn duration_sanity(diff: Duration) {
```
Comment (thomcc) [inline tests/pass/shims/time-with-isolation.rs:5]: nit: It's unclear from this name what this function means (additionally, it's not exactly great to use terms like sanity in code, imo). It might be better just inlined into the caller.

Comment (pvdrz) [reply tests/pass/shims/time-with-isolation.rs:5]: This file is a copy of `test.rs` minus the parts that use `SystemTime`.  I don't have a strong opinion about the name of the function but I'd change both if that was the case.

File: src/shims/time.rs [hunk 2]
```diff
@@ -1,8 +1,68 @@
...
+    }
+
+    pub fn checked_add_since_now(&self, duration: Duration) -> Option<Time> {
+        match self {
+            Self::Host(_) => Instant::now().checked_add(duration).map(Time::Monotonic),
+            Self::Virtual(nanoseconds) =>
+                nanoseconds
+                    .load(Ordering::Relaxed)
+                    .checked_add(duration.as_nanos().try_into().unwrap())
+                    .map(|duration| Time::Virtual(duration, nanoseconds.clone())),
```
Comment (oli-obk) [inline src/shims/time.rs:53]: why does this clone the `Arc` instead of taking a snapshot? Is that time meant to be stored somewhere across steps?

Comment (oli-obk) [reply src/shims/time.rs:53]: from what I can see we'll want a snapshot here instead of the updating time, all other `Time` things are snapshots after all.

Comment (pvdrz) [reply src/shims/time.rs:53]: I think this confusion is due to poor naming from my part. To know how much time remains before `Time` happens you need both the instant when time is due (first field) and the current instant (second field).

Comment (oli-obk) [reply src/shims/time.rs:53]: oh 🤦 the other enum variants can just read the global time, but here we need the "global virtual time".

Comment (oli-obk) [reply src/shims/time.rs:53]: document this in the enum please, that should avoid confusion in the future

File: src/concurrency/thread.rs [hunk 3]
```diff
@@ -184,6 +188,12 @@ impl<'mir, 'tcx> Thread<'mir, 'tcx> {
 #[derive(Debug)]
 pub enum Time {
     Monotonic(Instant),
+    Virtual {
+        /// The instant for this moment.
+        instant: u64,
+        /// A reference to the time anchor used to read the current global virtual time.
+        time_anchor: Arc<AtomicU64>,
```
Comment (RalfJung) [inline src/concurrency/thread.rs:195]: Why does this need an anchor when the others do not?

I was honestly thinking this use `Monotonic`. Only the code that actually tests whether an `Instant` has passed, and that sleeps until an `Instant` passes, should need adjusting.

File: src/machine.rs [hunk 4]
```diff
@@ -294,9 +293,6 @@ pub struct Evaluator<'mir, 'tcx> {
     /// The table of directory descriptors.
     pub(crate) dir_handler: shims::unix::DirHandler,
 
-    /// The "time anchor" for this machine's monotone clock (for `Instant` simulation).
-    pub(crate) time_anchor: Instant,
```
Comment (RalfJung) [inline src/machine.rs:298]: Why did you move this to `thread`? We use to have monotone clock support before we had threads. That's a separate subsystem.

File: src/shims/time.rs [hunk 5]
```diff
@@ -1,8 +1,71 @@
...
 
+use rustc_data_structures::sync::Ordering;
+
 use crate::concurrency::thread::Time;
 use crate::*;
 
+#[derive(Debug)]
+pub enum TimeAnchor {
+    Host(Instant),
+    Virtual(Arc<AtomicU64>),
```
Comment (RalfJung) [inline src/shims/time.rs:13]: Why is there an `Arc`? There should be a comment somewhere explaining the design. But I don't think an `Arc` should be needed.

File: src/shims/time.rs [hunk 6]
```diff
@@ -1,8 +1,71 @@
...
+}
+
+impl TimeAnchor {
+    pub fn new(communicate: bool) -> Self {
+        if communicate { Self::Host(Instant::now()) } else { Self::Virtual(Arc::new(0.into())) }
+    }
+
+    pub fn get(&self) -> Duration {
```
Comment (RalfJung) [inline src/shims/time.rs:21]: Please add doc comments to all these functions.

File: src/shims/time.rs [hunk 7]
```diff
@@ -1,8 +1,71 @@
...
+            Self::Host(instant) => Instant::now().saturating_duration_since(*instant),
+            Self::Virtual(nanoseconds) => Duration::from_nanos(nanoseconds.load(Ordering::Relaxed)),
+        }
+    }
+
+    pub fn tick(&self) {
+        match self {
+            Self::Host(_) => (),
+            Self::Virtual(nanoseconds) => {
+                nanoseconds.fetch_add(1, Ordering::Relaxed);
```
Comment (RalfJung) [inline src/shims/time.rs:32]: One nanosecond per basic block? Wow that's a fast machine.^^

Please move this into a constant somewhere, like `NANOSECOND_PER_BASIC_BLOCK` or so, so that it is clear what this means. Also maybe 10ns would be a bit more realistic? Not that it really matters...

File: src/shims/time.rs [hunk 8]
```diff
@@ -1,8 +1,71 @@
...
+    pub fn get(&self) -> Duration {
+        match self {
+            Self::Host(instant) => Instant::now().saturating_duration_since(*instant),
+            Self::Virtual(nanoseconds) => Duration::from_nanos(nanoseconds.load(Ordering::Relaxed)),
+        }
+    }
+
+    pub fn tick(&self) {
+        match self {
+            Self::Host(_) => (),
```
Comment (RalfJung) [inline src/shims/time.rs:30]:
```suggestion
            Self::Host(_) => {
                // Time will pass without us doing anything.
            }
```

File: src/shims/time.rs [hunk 9]
```diff
@@ -1,8 +1,71 @@
...
+                nanoseconds.fetch_add(1, Ordering::Relaxed);
+            }
+        }
+    }
+
+    pub fn sleep(&self, duration: Duration) {
+        match self {
+            Self::Host(_) => std::thread::sleep(duration),
+            Self::Virtual(nanoseconds) => {
+                nanoseconds.fetch_add(duration.as_nanos().try_into().unwrap(), Ordering::Relaxed);
```
Comment (RalfJung) [inline src/shims/time.rs:41]:
```suggestion
                // Just pretend that we have slept for some time.
                nanoseconds.fetch_add(duration.as_nanos().try_into().unwrap(), Ordering::Relaxed);
```

File: src/shims/time.rs [hunk 10]
```diff
@@ -1,8 +1,71 @@
...
+            }
+        }
+    }
+
+    pub fn checked_add_since_now(&self, duration: Duration) -> Option<Time> {
```
Comment (RalfJung) [inline src/shims/time.rs:46]: I don't know what these functions are even supposed to do. Seems related to why we have an `Arc`. But I can't figure out why...

File: src/shims/time.rs [hunk 11]
```diff
@@ -127,7 +190,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
         // QueryPerformanceCounter uses a hardware counter as its basis.
         // Miri will emulate a counter with a resolution of 1 nanosecond.
-        let duration = Instant::now().duration_since(this.machine.time_anchor);
+        let duration = this.machine.threads.time_anchor.get();
```
Comment (RalfJung) [inline src/shims/time.rs:193]: There's a `check_no_isolation` above that seems wrong now.
Why did the tests not catch that?

File: tests/pass/shims/time-with-isolation.rs [hunk 12]
```diff
@@ -0,0 +1,34 @@
+//@compile-flags: -Zmiri-disable-isolation
```
Comment (RalfJung) [inline tests/pass/shims/time-with-isolation.rs:1]: The test is called "with isolation" and here you disable isolation... :thinking:

File: src/concurrency/thread.rs [hunk 13]
```diff
@@ -192,6 +202,8 @@ impl Time {
     fn get_wait_time(&self) -> Duration {
```
Comment (RalfJung) [inline src/concurrency/thread.rs:202]: You'll probably have to pass the `Machine` to this function so that it can access the virtual clock to compute the wait time.

Comment (pvdrz) [reply src/concurrency/thread.rs:202]: Yeah I thought it would be better to keep `get_wait_time` signature unaltered and keep a reference to the global clock in the `Time`. I assumed borrowing issues would appear by having to pass a reference the machine or part of it as an argument but I can try and see what happens.

File: tests/pass/shims/time-with-isolation.rs (hunk 12 above)
Comment (pvdrz) [reply tests/pass/shims/time-with-isolation.rs:1]: Whoops

File: src/shims/time.rs (hunk 8 above)
Comment (pvdrz) [reply src/shims/time.rs:30]: Is this a joke about our windows support :rofl:?

File: src/shims/time.rs (hunk 7 above)
Comment (pvdrz) [reply src/shims/time.rs:32]: > One nanosecond per basic block? Wow that's a fast machine.^^

You wanted miri to be the fastest, didn't you?

> Please move this into a constant somewhere, like `NANOSECOND_PER_BASIC_BLOCK` or so, so that it is clear what this means. Also maybe 10ns would be a bit more realistic? Not that it really matters...

Sure I'll do the change

File: src/machine.rs (hunk 4 above)
Comment (pvdrz) [reply src/machine.rs:298]: because you need access to the global virtual clock inside `thread`. If we keep the `Arc` in `Time` then we need the `time_anchor` to clone it. If we remove the `Arc` then we need the `time_anchor` to call `get_wait_time`.

File: src/concurrency/thread.rs (hunk 3 above)
Comment (pvdrz) [reply src/concurrency/thread.rs:195]: Is just because you need to access the time_anchor to compute how much time is left when calling `get_wait_time`. 

Maybe I misunderstood but were you thinking about manually increasing `Instant` by `10 ns` on every tick and use that as the anchor?

File: src/shims/time.rs (hunk 9 above)
Comment (pvdrz) [reply src/shims/time.rs:41]: Is this a joke about our sleeping habits?

File: src/shims/time.rs (hunk 10 above)
Comment (pvdrz) [reply src/shims/time.rs:46]: This one computes a `Time` value corresponding to `now + duration`. I'd be happy to receive suggestions for this name :p 

Edit: Yes this exists because of the `Arc` but even if we didn't have it, we would have to pass the virtual clock here somehow.

File: tests/pass/shims/time-with-isolation.rs [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    for _ in 0..10 {
+        drop(vec![42]);
+    }
+    let now2 = Instant::now();
+    assert!(now2 > now1);
+    // Sanity-check the difference we got.
+    let diff = now2.duration_since(now1);
+    assert_eq!(now1 + diff, now2);
+    assert_eq!(now2 - diff, now1);
+    assert_eq!(diff.as_micros());
```
Comment (pvdrz) [inline tests/pass/shims/time-with-isolation.rs:31]:
```suggestion
    duration_sanity(diff);
```

File: src/concurrency/thread.rs (hunk 13 above)
Comment (RalfJung) [reply src/concurrency/thread.rs]: I'd prefer explicitly borrowing here over implicitly sharing an `Arc` in a bunch of hard-to-track places.

File: src/concurrency/thread.rs [hunk 15]
```diff
@@ -188,22 +186,17 @@ impl<'mir, 'tcx> Thread<'mir, 'tcx> {
...
 impl Time {
     /// How long do we have to wait from now until the specified time?
-    fn get_wait_time(&self) -> Duration {
+    fn get_wait_time(&self, clock: &Clock) -> Duration {
         match self {
             Time::Monotonic(instant) => instant.saturating_duration_since(Instant::now()),
-            Time::Virtual { instant, time_anchor } =>
-                Duration::from_nanos(instant - time_anchor.load(Ordering::Relaxed)),
+            Time::Virtual { nanoseconds } =>
+                Duration::from_nanos(nanoseconds - clock.assert_virtual().load(Ordering::Relaxed)),
```
Comment (RalfJung) [inline src/concurrency/thread.rs:199]: Looks like `Clock` should have a `get_wait_time` method abstracting this -- then we don't need to expose the underlying clock value ever.

File: src/shims/time.rs [hunk 16]
```diff
@@ -1,8 +1,99 @@
+use std::sync::atomic::AtomicU64;
 use std::time::{Duration, Instant, SystemTime};
 
+use rustc_data_structures::sync::Ordering;
+
 use crate::concurrency::thread::Time;
 use crate::*;
 
+const NANOSECOND_PER_BASIC_BLOCK: u64 = 10;
```
Comment (RalfJung) [inline src/shims/time.rs:9]:
```suggestion
/// When using a virtual clock, this defines how many nanoseconds do we pretend
/// are passing for each basic block.
const NANOSECOND_PER_BASIC_BLOCK: u64 = 10;
```

File: tests/pass/shims/time-with-isolation.rs [hunk 17]
```diff
@@ -26,7 +20,9 @@ fn main() {
     let diff = now2.duration_since(now1);
     assert_eq!(now1 + diff, now2);
     assert_eq!(now2 - diff, now1);
-    duration_sanity(diff);
+    // The virtual clock is deterministic and I got 29us. This should be updated if
+    // `NANOSECONDS_PER_BASIC_BLOCK` changes.
+    assert_eq!(diff.as_micros(), 29);
```
Comment (RalfJung) [inline tests/pass/shims/time-with-isolation.rs]: Might be good to add some slack; this depends on exactly how many basic blocks there are in `drop(vec![42])`, which can change as rustc changes.

File: src/concurrency/thread.rs [hunk 18]
```diff
@@ -184,14 +186,17 @@ impl<'mir, 'tcx> Thread<'mir, 'tcx> {
...
 }
 
 impl Time {
     /// How long do we have to wait from now until the specified time?
-    fn get_wait_time(&self) -> Duration {
+    fn get_wait_time(&self, clock: &Clock) -> Duration {
         match self {
             Time::Monotonic(instant) => instant.saturating_duration_since(Instant::now()),
+            Time::Virtual { nanoseconds } =>
+                Duration::from_nanos(nanoseconds - clock.assert_virtual().load(Ordering::Relaxed)),
```
Comment (RalfJung) [inline src/concurrency/thread.rs:199]: All of the pattern matching between `Monotonic` and `Virtual` should be isolated inside `time.rs`. I think this can be achieved by having that file just export a type `Instant` and using that in the rest of Miri.

I think you can make that type just be a simple `u64` (nanoseconds), interpreted relative to the time_anchor for 'real' monotone time. But you can also make it an `enum` that is either a `std::time::Instant` or a `u64`, I don't have a strong preference either way. But it should be private to `shims/time.rs`.

File: src/shims/time.rs [hunk 19]
```diff
@@ -1,8 +1,99 @@
...
+            Self::Host { .. } => std::thread::sleep(duration),
+            Self::Virtual { nanoseconds } => {
+                // Just pretend that we have slept for some time.
+                nanoseconds.fetch_add(duration.as_nanos().try_into().unwrap(), Ordering::Relaxed);
+            }
+        }
+    }
+
+    /// Compute `now + duration` relative to this clock.
+    pub fn checked_add_since_now(&self, duration: Duration) -> Option<Time> {
```
Comment (RalfJung) [inline src/shims/time.rs:67]: Oh now I get it. I would suggest `get_time_relative` as name for this method, and `get_time_absolute` for the other one.

File: src/shims/time.rs [hunk 20]
```diff
@@ -1,8 +1,99 @@
...
+    pub fn checked_add_since_start(&self, duration: Duration) -> Option<Time> {
+        match self {
+            Self::Host { time_anchor } => time_anchor.checked_add(duration).map(Time::Monotonic),
+            Self::Virtual { .. } =>
+                Some(Time::Virtual { nanoseconds: duration.as_nanos().try_into().unwrap() }),
+        }
+    }
+
+    /// Assert that this clock is a virtual one and get the current time in nanoseconds.
+    pub(crate) fn assert_virtual(&self) -> &AtomicU64 {
```
Comment (RalfJung) [inline src/shims/time.rs:89]: This should not be public, it should be hidden inside this file (if it is needed at all).

File: src/shims/time.rs [hunk 21]
```diff
@@ -1,8 +1,145 @@
...
+            ClockKind::Host { time_anchor: StdInstant::now() }
+        } else {
+            ClockKind::Virtual { nanoseconds: 0.into() }
+        };
+
+        Self { kind }
+    }
+
+    /// Get the current time relative to this clock.
+    pub fn get(&self) -> Duration {
```
Comment (RalfJung) [inline src/shims/time.rs:55]: Now we have `get` and `get_time_relative/absolute` and they have quite different types... that's somewhat confusing. Hm...

**Comment by RalfJung** [comment]
I did some tweaking to the API to make it match the one from the standard library more closely, which overall makes this a much smaller conceptual change for the rest of Miri. Please take a look and let me know what you think. :)

If you like it, please squash the commits together a little -- this is ready to go, as far as I am concerned.

File: src/clock.rs [hunk 22]
```diff
@@ -0,0 +1,119 @@
+use std::sync::atomic::AtomicU64;
+use std::time::{Duration, Instant as StdInstant};
+
+use rustc_data_structures::sync::Ordering;
```
Comment (saethlin) [inline src/clock.rs:4]: That's a funny place to import this from. We should import this from `std::sync::atomic`. Does rustc suggest this import?

File: src/clock.rs [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+
+use crate::*;
+
+/// When using a virtual clock, this defines how many nanoseconds do we pretend
+/// are passing for each basic block.
+const NANOSECOND_PER_BASIC_BLOCK: u64 = 10;
```
Comment (saethlin) [inline src/clock.rs:10]:
```suggestion
const NANOSECONDS_PER_BASIC_BLOCK: u64 = 10;
```

File: src/shims/time.rs [hunk 24]
```diff
@@ -40,9 +38,10 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             [this.eval_libc_i32("CLOCK_MONOTONIC")?, this.eval_libc_i32("CLOCK_MONOTONIC_COARSE")?];
 
         let duration = if absolute_clocks.contains(&clk_id) {
+            this.check_no_isolation("`clock_gettime` with real time clocks")?;
```
Comment (saethlin) [inline src/shims/time.rs:41]:
```suggestion
            this.check_no_isolation("`clock_gettime` with REALTIME clocks")?;
```
The use of lowercase "real" here is a bit ambiguious because there is the real-ness of the argument to `clock_gettime` and the real-ness of Miri's clock.

File: tests/pass/shims/time-with-isolation.rs [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+
+fn test_sleep() {
+    // We sleep a *long* time here -- but the clock is virtual so the test should still pass quickly.
+    let before = Instant::now();
+    std::thread::sleep(Duration::from_secs(3600));
+    let after = Instant::now();
+    assert!((after - before).as_secs() >= 3600);
+}
+
+/// Ensure that time passes even if we don't sleep (but just wor).
```
Comment (saethlin) [inline tests/pass/shims/time-with-isolation.rs:11]: `wor`?

File: src/clock.rs [hunk 26]
```diff
@@ -0,0 +1,119 @@
+use std::sync::atomic::AtomicU64;
+use std::time::{Duration, Instant as StdInstant};
+
+use rustc_data_structures::sync::Ordering;
+
+use crate::*;
+
+/// When using a virtual clock, this defines how many nanoseconds do we pretend
```
Comment (saethlin) [inline src/clock.rs:8]:
```suggestion
/// When using a virtual clock, this defines how many nanoseconds we pretend
```

**Review by saethlin — DISMISSED** [review]
I just have nits. This is a much smaller and more maintainable change than I somehow expected!

File: src/clock.rs (hunk 22 above)
Comment (pvdrz) [reply src/clock.rs:4]: indeed :p

**Comment by saethlin** [comment]
:+1: Squash the commits down some then I'll approve this. (also if you want a nice description of this to live in the commit history you could write a PR description)
